coral-incremental/src/main/java/com/linkedin/coral/incremental/RelNodeIncrementalTransformer.java

```diff
@@ -49,61 +66,130 @@ public RelNode visit(TableScan scan) {
+        IncrementalTransformerResults incrementalTransformerResultsChild = convertRelIncremental(project.getInput());
+        RelNode transformedChild = incrementalTransformerResultsChild.getIncrementalRelNode();
+        incrementalTransformerResults
+            .addMultipleIntermediateQueryRelNodes(incrementalTransformerResultsChild.getIntermediateQueryRelNodes());
+        incrementalTransformerResults.addIntermediateQueryRelNode(getTableNameFromDescription(project), project);
+        LogicalProject transformedProject =
+            LogicalProject.create(transformedChild, project.getProjects(), project.getRowType());
+        incrementalTransformerResults.addIntermediateQueryRelNode(getTableNameFromDescription(project) + "_delta",
+            transformedProject);
+        return transformedProject;
       }
 
       @Override
       public RelNode visit(LogicalUnion union) {
         List<RelNode> children = union.getInputs();
-        List<RelNode> transformedChildren =
+        List<IncrementalTransformerResults> incrementalTransformerResultsChildren =
             children.stream().map(child -> convertRelIncremental(child)).collect(Collectors.toList());
+        List<RelNode> transformedChildren = new ArrayList<>();
+        for (IncrementalTransformerResults incrementalTransformerResultsChild : incrementalTransformerResultsChildren) {
+          transformedChildren.add(incrementalTransformerResultsChild.getIncrementalRelNode());
+          incrementalTransformerResults
+              .addMultipleIntermediateQueryRelNodes(incrementalTransformerResultsChild.getIntermediateQueryRelNodes());
+        }
         return LogicalUnion.create(transformedChildren, union.all);
       }
 
       @Override
       public RelNode visit(LogicalAggregate aggregate) {
-        RelNode transformedChild = convertRelIncremental(aggregate.getInput());
+        IncrementalTransformerResults incrementalTransformerResultsChild = convertRelIncremental(aggregate.getInput());
+        RelNode transformedChild = incrementalTransformerResultsChild.getIncrementalRelNode();
+        incrementalTransformerResults
+            .addMultipleIntermediateQueryRelNodes(incrementalTransformerResultsChild.getIntermediateQueryRelNodes());
         return LogicalAggregate.create(transformedChild, aggregate.getGroupSet(), aggregate.getGroupSets(),
             aggregate.getAggCallList());
       }
     };
-    return originalNode.accept(converter);
+    incrementalTransformerResults.setIncrementalRelNode(originalNode.accept(converter));
+    return incrementalTransformerResults;
   }
 
-  private static LogicalProject createProjectOverJoin(LogicalJoin join, RelNode left, RelNode right,
+  private static String getTableNameFromDescription(RelNode relNode) {
+    String identifier = relNode.getDescription().split("#")[1];
+    return "Table#" + identifier;
+  }
+
+  private static LogicalProject createReplacementProjectNodeForGivenRelNode(String relOptTableName, RelNode relNode,
```

Comment:
Why is this needed? Hope it is not required. Any customization to the RelNodes means more fragility.

Reply:
This is needed to convert a join to a Project-TableScan for the table/view substitutions but I'm not sure that there is necessarily any RelNode customization here.

Reply:
Can you add a comment here to clarify what is the type of 'RelNode' passed here.